cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+	globalCommitment abstract.Point
+	// challenge holds the challenge for this round
+	challenge abstract.Secret
+	// response is our own computed response
+	response abstract.Secret
+	// aggregateResponses is the aggregated response from the children + our own
+	aggregateResponse abstract.Secret
+}
+
+// NewCosi returns a new Cosi struct given the suite, the longterm secret, and
+// the list of public keys. If some signers were not to be participating, you
+// have to set the mask using `SetMask` method. By default, all participants are
+// designated as participating.
+func NewCosi(suite abstract.Suite, private abstract.Secret, publics []abstract.Point) *Cosi {
+	cos := &Cosi{
+		suite:   suite,
+		private: private,
+		publics: publics,
+	}
+	// Start with an all-disabled participation mask, then set it correctly
+	cos.mask = NewCosiMask(suite, publics)
+	return cos
+}
+
+// Announcement holds only the timestamp for that round
+type Announcement struct {
+	Timestamp int64
+}
+
+// Commitment sends it's own commit Vi and the aggregate children's
+// commit V^i
+type Commitment struct {
+	Commitment     abstract.Point
+	ChildrenCommit abstract.Point
+}
+
+// Challenge is the Hash of Commit || Publics || Msg
+// Commit is the aggregated global commit
+type Challenge struct {
+	Challenge        abstract.Secret
+	GlobalCommitment abstract.Point
+}
+
+// Response holds the actual node's response ri and the
+// aggregate response r^i
+type Response struct {
+	Response     abstract.Secret
+	ChildrenResp abstract.Secret
+}
+
+// CreateAnnouncement creates an Announcement message with the timestamp set
+// to the current time.
+func (c *Cosi) CreateAnnouncement() *Announcement {
+	now := time.Now().Unix()
+	c.timestamp = now
+	return &Announcement{now}
+}
+
+// Announce stores the timestamp and relays the message.
+func (c *Cosi) Announce(in *Announcement) *Announcement {
+	c.timestamp = in.Timestamp
+	return in
+}
+
+// CreateCommitment creates the commitment out of the random secret and returns
+// the message to pass up in the tree. This is typically called by the leaves.
+func (c *Cosi) CreateCommitment(r io.Reader) *Commitment {
+	c.genCommit(r)
+	return &Commitment{
+		Commitment: c.commitment,
+	}
+}
+
+// Commit creates the commitment / secret + aggregate children commitments from
+// the children's messages.
+func (c *Cosi) Commit(r io.Reader, comms []*Commitment) *Commitment {
+	// generate our own commit
+	c.genCommit(r)
+
+	// take the children commitment
+	childVHat := c.suite.Point().Null()
+	for _, com := range comms {
+		// Add commitment of one child
+		childVHat = childVHat.Add(childVHat, com.Commitment)
+		// add commitment of it's children if there is one (i.e. if it is not a
+		// leaf)
+		if com.ChildrenCommit != nil {
+			childVHat = childVHat.Add(childVHat, com.ChildrenCommit)
+		}
+	}
+	// add our own commitment to the global V_hat
+	c.aggregateCommitment = c.suite.Point().Add(childVHat, c.commitment)
```

Comment:
Why not initialize childVHat to c.commitment (our local commitment) in the first place up above, rather than initializing it to null and then having to Add it separately later?